src/api/bulk2.ts

```diff
@@ -0,0 +1,1000 @@
+
+  constructor(connection: Connection<S>) {
+    this.connection = connection;
+    this.logger = connection._logger;
+    this.logger = this.connection._logLevel
+      ? getLogger('bulk2').createInstance(this.connection._logLevel)
+      : getLogger('bulk2');
+  }
+
+  /**
+   * Create an instance of an ingest job object.
+   *
+   * @params {NewIngestJobOptions} options object
+   * @returns {IngestJobV2} An ingest job instance
+   * @example
+   * // Upsert records to the Account object.
+   *
+   * const job = connection.bulk2.createJob({
+   *   operation: 'insert'
+   *   object: 'Account',
+   * });
+   *
+   * // create the job in the org
+   * await job.open()
+   *
+   * // upload data
+   * await job.uploadData(csvFile)
+   *
+   * // finished uploading data, mark it as ready for processing
+   * await job.close()
+   */
+  public createJob<Opr extends IngestOperation>(
+    options: NewIngestJobOptions,
+  ): IngestJobV2<S, Opr> {
+    return new IngestJobV2({
+      connection: this.connection,
+      jobInfo: options,
+      pollingOptions: this,
+    });
+  }
+
+  /**
+   * Get an ingest job instance specified by a given job ID
+   *
+   * @param options Options object with a job ID
+   * @returns IngestJobV2 An ingest job
+   */
+  public job<Opr extends IngestOperation>(
+    options: ExistingIngestJobOptions,
+  ): IngestJobV2<S, Opr> {
+    return new IngestJobV2({
+      connection: this.connection,
+      jobInfo: options,
+      pollingOptions: this,
+    });
+  }
+
+  /**
+   * Create, upload, and start bulkload job
+   */
+  async loadAndWaitForResults(
+    options: NewIngestJobOptions &
+      Partial<BulkV2PollingOptions> & {
+        input: Record[] | Readable | string;
+      },
+  ): Promise<IngestJobV2Results<S>> {
+    if (!options.pollTimeout) options.pollTimeout = this.pollTimeout;
+    if (!options.pollInterval) options.pollInterval = this.pollInterval;
+
+    const job = this.createJob(options);
+    try {
+      await job.open();
+      await job.uploadData(options.input);
+      await job.close();
+      await job.poll(options.pollInterval, options.pollTimeout);
+      return await job.getAllResults();
+    } catch (error) {
+      const err = error as Error;
+      if (err.name !== 'JobPollingTimeoutError') {
+        // fires off one last attempt to clean up and ignores the result | error
+        job.delete().catch((ignored) => ignored);
+      }
+      throw err;
+    }
+  }
+
+  /**
+   * Execute bulk query and get a record stream.
+   *
+   * Default timeout: 10000ms
+   *
+   * @param soql SOQL query
+   * @param BulkV2PollingOptions options object
+   *
+   * @returns {RecordStream} - Record stream, convertible to a CSV data stream
+   */
+  async query(
+    soql: string,
+    options?: Partial<BulkV2PollingOptions> & {
+      scanAll?: boolean;
+      columnDelimiter?: QueryJobInfoV2['columnDelimiter'];
+      lineEnding?: QueryJobInfoV2['lineEnding'];
+    },
+  ): Promise<Parsable<Record>> {
```

Comment:
`bulk2.query` now returns a duplex, consumers should listen to the record event and handle records there:
https://jsforce.github.io/document/#bulk-query

Unlike `conn.bulk`, you need to `await conn.bulk2.query`.

`conn.bulk` has an async IIFE: https://github.com/jsforce/jsforce/blob/588c6cf986650fb68f9b41314d5c48a9a5c5d478/src/api/bulk.ts#L969, bulk2 doesn't.

Check the e2e test below to see how it collects all records from the stream.

Reply:
sounds good. There are a few async methods there, so they would match bulk2 👍🏼